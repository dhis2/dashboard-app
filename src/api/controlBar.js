import { getInstance } from 'd2/lib/d2';
import { DEFAULT_ROWS } from '../reducers/controlBar';

const NAMESPACE = 'dashboard';
const KEY = 'controlBarRows';

const hasNamespace = async d2 => await d2.currentUser.dataStore.has(NAMESPACE);

const getNamespace = async (d2, hasNamespace) =>
    hasNamespace
        ? await d2.currentUser.dataStore.get(NAMESPACE)
        : await d2.currentUser.dataStore.create(NAMESPACE);

export const apiGetControlBarRows = async () => {
    const d2 = await getInstance();
    const namespace = await getNamespace(d2, await hasNamespace(d2));
    const hasKey = namespace.keys && namespace.keys.find(key => key === KEY);

    if (hasKey) {
        return await namespace.get(KEY);
    } else {
        await apiPostControlBarRows(DEFAULT_ROWS, namespace);
        console.log(
            '(These errors to /userDataStore/dashboard can be ignored)'
        );
        return DEFAULT_ROWS;
    }
};

export const apiPostControlBarRows = async (rows, namespace) => {
    const d2 = await getInstance();
    const ns = namespace || (await getNamespace(d2, hasNamespace));

    ns.set(KEY, rows);
};
