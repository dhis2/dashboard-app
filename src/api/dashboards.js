import { getInstance } from 'd2/lib/d2';
import arrayClean from 'd2-utilizr/lib/arrayClean';
import {
    onError,
    getEndPointName,
    getDashboardFields,
    getFavoriteFields,
} from './index';

// Get "all" dashboards on startup
export const apiFetchDashboards = () =>
    getInstance()
        .then(d2 =>
            d2.models.dashboard.list({
                fields: [
                    getDashboardFields().join(','),
                    'dashboardItems[id]',
                ].join(','),
                paging: 'false',
            })
        )
        .catch(onError);

// Get more info about selected dashboard
export const apiFetchSelected = id =>
    getInstance()
        .then(d2 =>
            d2.models.dashboard.get(id, {
                fields: arrayClean(
                    getDashboardFields({
                        withItems: true,
                        withFavorite: { withDimensions: false },
                    })
                ).join(','),
            })
        )
        .catch(onError);

// Get more info about the favorite of a dashboard item
export const apiFetchFavorite = (id, type) =>
    getInstance().then(d2 =>
        d2.Api.getApi().get(
            `${getEndPointName(type)}/${id}?fields=${getFavoriteFields({
                withDimensions: true,
                withOptions: true,
            })}`
        )
    );

// Star dashboard
export const apiStarDashboard = (id, isStarred) => {
    const url = `dashboards/${id}/favorite`;

    getInstance().then(d2 => {
        if (isStarred) {
            d2.Api.getApi().post(url);
        } else {
            d2.Api.getApi().delete(url);
        }
    });
};
