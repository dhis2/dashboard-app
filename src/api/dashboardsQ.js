export const getDashboardsQQuery = (query = '', maxItems = []) => {
    return {
        resource: `dashboards/q/${query}`,
        params: { count: 11, max: maxItems },
    }
}
