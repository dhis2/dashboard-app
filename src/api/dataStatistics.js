import { getInstance } from 'd2';

export const apiPostDataStatistics = async (eventType, id) => {
    const d2 = await getInstance();
    const url = `dataStatistics?eventType=${eventType}&favorite=${id}`;

    d2.Api.getApi().post(url);
};
