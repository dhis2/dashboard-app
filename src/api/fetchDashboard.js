import arrayClean from 'd2-utilizr/lib/arrayClean'
import { isViewMode } from '../modules/dashboardModes'
import { getCustomDashboards } from '../modules/getCustomDashboards'
import { withShape } from '../modules/gridUtil'
import {
    getIdNameFields,
    getListItemFields,
    getFavoritesFields,
} from './metadata'

const getDashboardItemsFields = () =>
    arrayClean([
        'id',
        'type',
        'shape',
        'x',
        'y',
        'width~rename(w)',
        'height~rename(h)',
        'messages',
        'text',
        'appKey',
        `${getListItemFields().join(',')}`,
        `${getFavoritesFields().join(',')}`,
    ])

const baseDashboardFields = arrayClean([
    'id',
    'displayName',
    'displayDescription',
    'favorite~rename(starred)',
    'access',
    'restrictFilters',
    'allowedFilters',
    'layout',
    'itemConfig',
    `dashboardItems[${getDashboardItemsFields().join(',')}]`,
])

export const viewDashboardQuery = {
    resource: 'dashboards',
    id: ({ id }) => id,
    params: {
        fields: baseDashboardFields,
    },
}

export const editDashboardQuery = {
    resource: 'dashboards',
    id: ({ id }) => id,
    params: {
        fields: arrayClean([
            ...baseDashboardFields,
            `user[${getIdNameFields({ rename: true }).join(',')}]`,
            'name',
            'description',
            'created',
            'favorite',
            'favorites',
            'lastUpdated',
            'href', // needed for d2-ui-translations-dialog
        ]),
    },
}

// Get more info about selected dashboard
export const apiFetchDashboard = async (
    dataEngine,
    id,
    { mode = null, forSave = false } = {}
) => {
    const query = isViewMode(mode) ? viewDashboardQuery : editDashboardQuery

    try {
        const dashboardData = await dataEngine.query(
            { dashboard: query },
            {
                variables: {
                    id,
                },
            }
        )

        const dashboard = dashboardData.dashboard

        if (!forSave) {
            return getCustomDashboards(
                Object.assign({}, dashboard, {
                    dashboardItems: withShape(dashboard.dashboardItems),
                })
            )[0]
        }
        return dashboard
    } catch (error) {
        console.log('Error: ', error)
    }
}
