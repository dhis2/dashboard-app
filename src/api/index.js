import { getInstance } from 'd2/lib/d2';
import arrayClean from 'd2-utilizr/lib/arrayClean';

import { favoriteTypeUrlMap } from '../util';

export const getUrlByFavoriteType = type =>
    favoriteTypeUrlMap[type].endPointName;

export const delay = (ms = 500) =>
    new Promise(resolve => setTimeout(resolve, ms));

const getIdNameFields = ({ rename } = {}) => [
    'id',
    `${rename ? 'displayName~rename(name)' : 'name'}`,
];

const getItemFields = () => ['dimensionItem~rename(id)'];

const getDimensionFields = ({ withItems }) =>
    arrayClean([
        'dimension',
        withItems ? `items[${getItemFields().join(',')}]` : ``,
    ]);

const getAxesFields = ({ withItems }) => [
    `columns[${getDimensionFields({ withItems }).join(',')}]`,
    `rows[${getDimensionFields({ withItems }).join(',')}]`,
    `filters[${getDimensionFields({ withItems }).join(',')}]`,
];

const interpretationFields = () => {
    return 'interpretations[id]';
    // return 'interpretations[id,text,created,user[id,displayName],likedBy,comments[id,text,created,user[id,displayName]]]';
};

const getFavoriteFields = ({ withDimensions, withOptions }) => {
    return arrayClean([
        `${getIdNameFields({ rename: true }).join(',')}`,
        interpretationFields(),
        withDimensions ? `${getAxesFields({ withItems: true }).join(',')}` : ``,
        withOptions
            ? [
                  '*',
                  '!attributeDimensions',
                  '!attributeValues',
                  '!category',
                  '!categoryDimensions',
                  '!categoryOptionGroupSetDimensions',
                  '!columnDimensions',
                  '!dataDimensionItems',
                  '!dataElementDimensions',
                  '!dataElementGroupSetDimensions',
                  '!filterDimensions',
                  '!itemOrganisationUnitGroups',
                  '!lastUpdatedBy',
                  '!organisationUnitGroupSetDimensions',
                  '!organisationUnitLevels',
                  '!organisationUnits',
                  '!programIndicatorDimensions',
                  '!relativePeriods',
                  '!reportParams',
                  '!rowDimensions',
                  '!series',
                  '!translations',
                  '!userOrganisationUnit',
                  '!userOrganisationUnitChildren',
                  '!userOrganisationUnitGrandChildren',
              ].join(',')
            : '',
    ]);
};

const getFavoritesFields = ({ withDimensions, withOptions }) => [
    `reportTable[${getFavoriteFields({ withDimensions }).join(',')}]`,
    `chart[${getFavoriteFields({ withDimensions }).join(',')}]`,
    `map[${getFavoriteFields({ withDimensions }).join(',')}]`,
    `eventReport[${getFavoriteFields({ withDimensions }).join(',')}]`,
    `eventChart[${getFavoriteFields({ withDimensions }).join(',')}]`,
];

const getDashboardItemsFields = ({ withFavorite } = {}) =>
    arrayClean([
        'id',
        'type',
        'shape',
        withFavorite
            ? `${getFavoritesFields({
                  withDimensions: withFavorite.withDimensions,
              }).join(',')}`
            : ``,
    ]);

const getDashboardFields = ({ withItems, withFavorite } = {}) =>
    arrayClean([
        `${getIdNameFields({ rename: true }).join(',')}`,
        'description',
        `user[${getIdNameFields({ rename: true }).join(',')}]`,
        'created',
        'lastUpdated',
        withItems
            ? `dashboardItems[${getDashboardItemsFields({
                  withFavorite,
              }).join(',')}]`
            : ``,
    ]);

const onError = error => console.log('Error: ', error);

// Get "all" dashboards on startup
export const apiFetchDashboards = () =>
    getInstance()
        .then(d2 =>
            d2.models.dashboard.list({
                fields: [
                    getDashboardFields().join(','),
                    'dashboardItems[id]',
                ].join(','),
            })
        )
        .catch(onError);

// Create a new dashboard
export const apiPostDashboard = (props = {}) =>
    getInstance().then(d2 => d2.Api.getApi().post('/dashboards', props));

// Get more info about selected dashboard
export const apiFetchSelected = id =>
    getInstance()
        .then(d2 =>
            d2.models.dashboard.get(id, {
                fields: arrayClean(
                    getDashboardFields({
                        withItems: true,
                        withFavorite: { withDimensions: false },
                    })
                ).join(','),
            })
        )
        .catch(onError);

// Get more info about selected dashboard
export const apiFetchFavorite = (id, type) =>
    getInstance().then(d2 =>
        d2.Api.getApi().get(
            `${getUrlByFavoriteType(type)}/${id}?fields=${getFavoriteFields({
                withDimensions: true,
                withOptions: true,
            })}`
        )
    );
