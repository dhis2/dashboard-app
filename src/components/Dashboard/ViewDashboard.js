import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import isEmpty from 'lodash/isEmpty';
import i18n from '@dhis2/d2-i18n';

import {
    sGetAllDashboards,
    sDashboardsIsFetching,
} from '../../reducers/dashboards';
import { sGetSelectedId } from '../../reducers/selected';
import DashboardsBar from '../ControlBar/DashboardsBar';
import DashboardVerticalOffset from './DashboardVerticalOffset';
import DashboardContent from './DashboardContent';
import NoContentMessage from '../../widgets/NoContentMessage';

export const Content = ({ hasDashboardContent, dashboardsIsEmpty }) => {
    const msg = dashboardsIsEmpty
        ? i18n.t(
              'No dashboards found. Use the + button to create a new dashboard.'
          )
        : i18n.t('Requested dashboard not found');

    return hasDashboardContent ? (
        <DashboardContent editMode={false} />
    ) : (
        <NoContentMessage text={msg} />
    );
};

export const ViewDashboard = ({ id, dashboardsIsEmpty, dashboardsLoaded }) => {
    const hasDashboardContent = id && !dashboardsIsEmpty;
    const contentNotReady = !dashboardsLoaded || id === null;

    return (
        <Fragment>
            <DashboardsBar />
            <DashboardVerticalOffset />
            <div className="dashboard-wrapper">
                {contentNotReady ? null : (
                    <Content
                        hasDashboardContent={hasDashboardContent}
                        dashboardsIsEmpty={dashboardsIsEmpty}
                    />
                )}
            </div>
        </Fragment>
    );
};

const mapStateToProps = state => {
    const dashboards = sGetAllDashboards(state);

    return {
        id: sGetSelectedId(state),
        dashboardsIsEmpty: isEmpty(dashboards),
        dashboardsLoaded: !sDashboardsIsFetching(state),
    };
};

export default connect(mapStateToProps)(ViewDashboard);
