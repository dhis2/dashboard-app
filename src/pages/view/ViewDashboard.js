import { useOnlineStatus } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { AlertStack, AlertBar } from '@dhis2/ui'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { useState, useEffect } from 'react'
import { connect } from 'react-redux'
import { acClearEditDashboard } from '../../actions/editDashboard'
import { acSetPassiveViewRegistered } from '../../actions/passiveViewRegistered'
import { acClearPrintDashboard } from '../../actions/printDashboard'
import {
    tSetSelectedDashboardById,
    tSetSelectedDashboardByIdOffline,
} from '../../actions/selected'
import { apiPostDataStatistics } from '../../api/dataStatistics'
import DashboardContainer from '../../components/DashboardContainer'
import LoadingMask from '../../components/LoadingMask'
import Notice from '../../components/Notice'
import { setHeaderbarVisible } from '../../modules/setHeaderbarVisible'
import { useCacheableSection } from '../../modules/useCacheableSection'
import { sGetDashboardById } from '../../reducers/dashboards'
import { sGetPassiveViewRegistered } from '../../reducers/passiveViewRegistered'
import { sGetSelectedId } from '../../reducers/selected'
import DashboardsBar from './DashboardsBar/DashboardsBar'
import FilterBar from './FilterBar/FilterBar'
import ItemGrid from './ItemGrid'
import classes from './styles/ViewDashboard.module.css'
import TitleBar from './TitleBar/TitleBar'

const ViewDashboard = props => {
    const [controlbarExpanded, setControlbarExpanded] = useState(false)
    const [loadingMessage, setLoadingMessage] = useState(null)
    const [loaded, setLoaded] = useState(false)
    const [loadFailed, setLoadFailed] = useState(false)
    const { online } = useOnlineStatus()
    const { isCached, recordingState } = useCacheableSection(props.requestedId)

    useEffect(() => {
        setHeaderbarVisible(true)
        props.clearEditDashboard()
        props.clearPrintDashboard()
    }, [])

    useEffect(() => {
        setLoaded(false)

        Array.from(
            document.getElementsByClassName('dashboard-scroll-container')
        ).forEach(container => {
            container.scroll(0, 0)
        })
    }, [props.requestedId])

    useEffect(() => {
        if (!props.passiveViewRegistered && online) {
            apiPostDataStatistics('PASSIVE_DASHBOARD_VIEW', props.requestedId)
                .then(() => {
                    props.registerPassiveView()
                })
                .catch(error => console.info(error))
        }
    }, [props.passiveViewRegistered])

    useEffect(() => {
        const loadDashboard = async () => {
            const alertTimeout = setTimeout(() => {
                if (props.requestedDashboardName) {
                    setLoadingMessage(
                        i18n.t('Loading dashboard â€“ {{name}}', {
                            name: props.requestedDashboardName,
                        })
                    )
                } else {
                    setLoadingMessage(i18n.t('Loading dashboard'))
                }
            }, 500)

            try {
                setLoaded(true)
                await props.fetchDashboard(props.requestedId, props.username)

                setLoadFailed(false)
                setLoadingMessage(null)
                clearTimeout(alertTimeout)
            } catch (e) {
                setLoaded(false)
                setLoadFailed(true)
                setLoadingMessage(null)
                clearTimeout(alertTimeout)
                props.setSelectedAsOffline(props.requestedId, props.username)
            }
        }

        const requestedIsAvailable = online || isCached
        const switchingDashboard = props.requestedId !== props.currentId
        if (
            requestedIsAvailable &&
            (recordingState === 'recording' || !loaded)
        ) {
            loadDashboard()
        } else if (!requestedIsAvailable && switchingDashboard) {
            setLoaded(false)
            props.setSelectedAsOffline(props.requestedId, props.username)
        }
    }, [props.requestedId, props.currentId, loaded, recordingState, online])

    const onExpandedChanged = expanded => setControlbarExpanded(expanded)

    const getContent = () => {
        if (
            !online &&
            !isCached &&
            (props.requestedId !== props.currentId || !loaded)
        ) {
            return (
                <Notice
                    title={i18n.t('Offline')}
                    message={i18n.t(
                        'This dashboard cannot be loaded while offline.'
                    )}
                />
            )
        }

        if (loadFailed) {
            return (
                <Notice
                    title={i18n.t('Load dashboard failed')}
                    message={i18n.t(
                        'This dashboard could not be loaded. Please try again later.'
                    )}
                />
            )
        }

        return props.requestedId !== props.currentId ? (
            <LoadingMask />
        ) : (
            <>
                <TitleBar />
                <FilterBar />
                <ItemGrid />
            </>
        )
    }

    return (
        <>
            <div
                className={cx(classes.container, 'dashboard-scroll-container')}
                data-test="outer-scroll-container"
            >
                <DashboardsBar
                    expanded={controlbarExpanded}
                    onExpandedChanged={onExpandedChanged}
                />
                <DashboardContainer covered={controlbarExpanded}>
                    {controlbarExpanded && (
                        <div
                            className={classes.cover}
                            onClick={() => setControlbarExpanded(false)}
                        />
                    )}
                    {getContent()}
                </DashboardContainer>
            </div>
            <AlertStack>
                {loadingMessage && (
                    <AlertBar
                        onHidden={() => setLoadingMessage(null)}
                        permanent
                    >
                        {loadingMessage}
                    </AlertBar>
                )}
            </AlertStack>
        </>
    )
}

ViewDashboard.propTypes = {
    clearEditDashboard: PropTypes.func,
    clearPrintDashboard: PropTypes.func,
    currentId: PropTypes.string,
    fetchDashboard: PropTypes.func,
    passiveViewRegistered: PropTypes.bool,
    registerPassiveView: PropTypes.func,
    requestedDashboardName: PropTypes.string,
    requestedId: PropTypes.string,
    setSelectedAsOffline: PropTypes.func,
    username: PropTypes.string,
}

const mapStateToProps = (state, ownProps) => {
    const dashboard = sGetDashboardById(state, ownProps.requestedId) || {}

    return {
        passiveViewRegistered: sGetPassiveViewRegistered(state),
        requestedDashboardName: dashboard.displayName || null,
        currentId: sGetSelectedId(state),
    }
}

export default connect(mapStateToProps, {
    clearEditDashboard: acClearEditDashboard,
    clearPrintDashboard: acClearPrintDashboard,
    registerPassiveView: acSetPassiveViewRegistered,
    fetchDashboard: tSetSelectedDashboardById,
    setSelectedAsOffline: tSetSelectedDashboardByIdOffline,
})(ViewDashboard)
