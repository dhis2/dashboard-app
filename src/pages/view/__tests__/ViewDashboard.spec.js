import { render } from '@testing-library/react'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { apiPostDataStatistics } from '../../../api/dataStatistics'
import { apiFetchDashboard } from '../../../api/fetchDashboard'
import ViewDashboard from '../ViewDashboard'

jest.mock('@dhis2/app-runtime-adapter-d2', () => ({
    useD2: () => ({
        d2: {
            currentUser: {
                username: 'rainbowDash',
                id: 'r3nb0d5h',
            },
        },
    }),
}))

jest.mock('@dhis2/app-runtime', () => ({
    useDhis2ConnectionStatus: jest.fn(() => ({ isConnected: true })),
    useCacheableSection: jest.fn(() => ({
        isCached: false,
        recordingState: 'default',
    })),
}))

jest.mock('../../../api/fetchDashboard')

jest.mock(
    '../../../components/DashboardsBar/DashboardsBar',
    () =>
        function MockDashboardsBar() {
            return <div>DashboardsBar</div>
        }
)

jest.mock(
    '../TitleBar/TitleBar',
    () =>
        function MockTitleBar() {
            return <div>TitleBar</div>
        }
)

jest.mock(
    '../FilterBar/FilterBar',
    () =>
        function MockFilterBar() {
            return <div>MockFilterBar</div>
        }
)

jest.mock(
    '../ItemGrid',
    () =>
        function MockItemGrid() {
            return <div>MockItemGrid</div>
        }
)

jest.mock('../../../api/dataStatistics', () => ({
    apiPostDataStatistics: jest.fn(() => {
        return new Promise(resolve => resolve(true))
    }),
}))

const dashboardId = 'rainbowdash'
const store = {
    dashboards: {
        [dashboardId]: {
            id: dashboardId,
            displayName: 'Rainbow Dash',
            starred: true,
        },
    },

    selected: {
        id: dashboardId,
    },
    passiveViewRegistered: false,
}

window.HTMLElement.prototype.scroll = function () {}

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

test('ViewDashboard renders dashboard', async () => {
    const promise = Promise.resolve({
        id: dashboardId,
        dashboardItems: [],
    })
    apiFetchDashboard.mockResolvedValue(promise)

    const { container } = render(
        <>
            <header />
            <Provider store={mockStore(store)}>
                <ViewDashboard requestedId={dashboardId} />
            </Provider>
        </>
    )

    await act(() => promise)
    expect(container).toMatchSnapshot()
    jest.clearAllMocks()
})

test('ViewDashboard does not post passive view to api if passive view has been registered', async () => {
    store.passiveViewRegistered = true
    const promise = Promise.resolve({
        id: dashboardId,
        dashboardItems: [],
    })
    apiFetchDashboard.mockResolvedValue(promise)

    render(
        <>
            <header />
            <Provider store={mockStore(store)}>
                <ViewDashboard id={dashboardId} />
            </Provider>
        </>
    )

    await act(() => promise)
    expect(apiPostDataStatistics).not.toHaveBeenCalled()
    jest.clearAllMocks()
})

test('ViewDashboard posts passive view to api if passive view has not been registered', async () => {
    store.passiveViewRegistered = false
    const promise = Promise.resolve({
        id: dashboardId,
        dashboardItems: [],
    })
    apiFetchDashboard.mockResolvedValue(promise)

    render(
        <>
            <header />
            <Provider store={mockStore(store)}>
                <ViewDashboard id={dashboardId} />
            </Provider>
        </>
    )

    await act(() => promise)

    expect(apiPostDataStatistics).toHaveBeenCalled()
    jest.clearAllMocks()
})
