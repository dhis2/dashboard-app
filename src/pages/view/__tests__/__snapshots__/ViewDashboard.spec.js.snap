// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewDashboard renders dashboard 1`] = `
<div>
  <header />
  <div
    class="container dashboard-scroll-container"
    data-test="outer-scroll-container"
  >
    <div>
      DashboardsBar
    </div>
    <div
      class="container dashboard-scroll-container"
      data-test="inner-scroll-container"
    >
      <div>
        MockFilterBar
      </div>
      <div>
        MockItemGrid
      </div>
    </div>
  </div>
</div>
`;
